http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/WsmApiClientProvider.scala

```diff
@@ -11,27 +17,32 @@ import org.http4s.Uri
  * Based on `org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/WorkspaceManagerApiClientProvider.scala`
  *
  */
-trait WsmApiClientProvider {
-  def getControlledAzureResourceApi(token: String): ControlledAzureResourceApi
-  def getResourceApi(token: String): ResourceApi
+trait WsmApiClientProvider[F[_]] {
+  def getControlledAzureResourceApi(token: String)(implicit ev: Ask[F, AppContext]): F[ControlledAzureResourceApi]
+  def getResourceApi(token: String)(implicit ev: Ask[F, AppContext]): F[ResourceApi]
 }
 
-class HttpWsmClientProvider(baseWorkspaceManagerUrl: Uri) extends WsmApiClientProvider {
-  private def getApiClient: ApiClient = {
-    val client: ApiClient = new ApiClient()
-    client.setBasePath(baseWorkspaceManagerUrl.renderString)
-    client
-  }
+class HttpWsmClientProvider[F[_]](baseWorkspaceManagerUrl: Uri)(implicit F: Async[F]) extends WsmApiClientProvider[F] {
+  private def getApiClient(token: String)(implicit ev: Ask[F, AppContext]): F[ApiClient] =
+    for {
+      ctx <- ev.ask
+      client = new ApiClient() {
+        override def performAdditionalClientConfiguration(clientConfig: ClientConfig): Unit = {
```

Comment:
I am not certain this is thread safe. `performAdditionalClientConfiguration` is called during `ApiClient` construction. The constructed client is wrapped in a `F` which is then chained to other calls. Those other calls are not guaranteed to be on the same thread.

Reply:
Fascinating… maybe Leo should do something like this then? https://github.com/broadinstitute/rawls/blob/develop/core/src/main/scala/org/broadinstitute/dsde/rawls/util/WithSpanFilter.scala

Reply:
…I gave that a shot